From c6f0f7828de26ac2d68b56a7abe6f948befb0a6e Mon Sep 17 00:00:00 2001
From: Kevin Eyssartier <kevin.eyssartier@thalesgroup.com>
Date: Tue, 16 Nov 2021 23:47:15 +0100
Subject: [PATCH 01/15] cva6: Add Genesys 2 support (#6)

* cva6_genesysII board support

* Update cva6.c

Removing unecessary print.

* bootm: Work around lmb-.bss collision

lmb_alloc_base returns an address that collides with the .bss section.
As a result, memmove of the fdt overwrites some global variables such
as the console device pointer, completely breaking the system.

This workaround hardcodes the fdt's target address to a safe value.

Signed-off-by: Nils Wistoff <nwistoff@iis.ee.ethz.ch>

Co-authored-by: Nils Wistoff <nwistoff@iis.ee.ethz.ch>
---
 arch/riscv/Kconfig                           |   4 +
 arch/riscv/dts/Makefile                      |   1 +
 arch/riscv/dts/cva6_genesysII.dts            | 145 +++++++++++++++++++
 board/openhwgroup/cva6_genesysII/Kconfig     |  27 ++++
 board/openhwgroup/cva6_genesysII/Makefile    |   1 +
 board/openhwgroup/cva6_genesysII/cva6.c      |   9 ++
 common/image-fdt.c                           |   7 +
 configs/openhwgroup_cva6_genesysII_defconfig |  37 +++++
 include/configs/openhwgroup_cva6_genesysII.h |  16 ++
 9 files changed, 247 insertions(+)
 create mode 100644 arch/riscv/dts/cva6_genesysII.dts
 create mode 100644 board/openhwgroup/cva6_genesysII/Kconfig
 create mode 100644 board/openhwgroup/cva6_genesysII/Makefile
 create mode 100644 board/openhwgroup/cva6_genesysII/cva6.c
 create mode 100644 configs/openhwgroup_cva6_genesysII_defconfig
 create mode 100644 include/configs/openhwgroup_cva6_genesysII.h

diff --git a/arch/riscv/Kconfig b/arch/riscv/Kconfig
index b3d7fd84ce..16dac7a62a 100644
--- a/arch/riscv/Kconfig
+++ b/arch/riscv/Kconfig
@@ -23,6 +23,9 @@ config TARGET_SIFIVE_UNLEASHED
 config TARGET_SIFIVE_UNMATCHED
 	bool "Support SiFive Unmatched Board"
 
+config TARGET_OPENHWGROUP_CVA6_GENESYSII
+	bool "Target cva6 architecture on genesysII board"
+
 config TARGET_SIPEED_MAIX
 	bool "Support Sipeed Maix Board"
 
@@ -61,6 +64,7 @@ source "board/microchip/mpfs_icicle/Kconfig"
 source "board/sifive/unleashed/Kconfig"
 source "board/sifive/unmatched/Kconfig"
 source "board/sipeed/maix/Kconfig"
+source "board/openhwgroup/cva6_genesysII/Kconfig"
 
 # platform-specific options below
 source "arch/riscv/cpu/ax25/Kconfig"
diff --git a/arch/riscv/dts/Makefile b/arch/riscv/dts/Makefile
index 7778874831..2744f5948c 100644
--- a/arch/riscv/dts/Makefile
+++ b/arch/riscv/dts/Makefile
@@ -6,6 +6,7 @@ dtb-$(CONFIG_TARGET_QEMU_VIRT) += qemu-virt.dtb
 dtb-$(CONFIG_TARGET_SIFIVE_UNLEASHED) += hifive-unleashed-a00.dtb
 dtb-$(CONFIG_TARGET_SIFIVE_UNMATCHED) += hifive-unmatched-a00.dtb
 dtb-$(CONFIG_TARGET_SIPEED_MAIX) += k210-maix-bit.dtb
+dtb-$(CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII) += cva6_genesysII.dtb
 
 targets += $(dtb-y)
 
diff --git a/arch/riscv/dts/cva6_genesysII.dts b/arch/riscv/dts/cva6_genesysII.dts
new file mode 100644
index 0000000000..b699c7bbb0
--- /dev/null
+++ b/arch/riscv/dts/cva6_genesysII.dts
@@ -0,0 +1,145 @@
+/dts-v1/;
+
+/ {
+  #address-cells = <2>;
+  #size-cells = <2>;
+  compatible = "eth,ariane-bare-dev";
+  model = "eth,ariane-bare";
+  chosen {
+    stdout-path = "/soc/uart@10000000:115200";
+    //tick-timer =  "/cpus/cpu@0";
+  };
+  cpus {
+    #address-cells = <1>;
+    #size-cells = <0>;
+    timebase-frequency = <25000000>; // 25 MHz
+    CPU0: cpu@0 {
+      clock-frequency = <50000000>; // 50 MHz
+      device_type = "cpu";
+      reg = <0>;
+      status = "okay";
+      compatible = "eth, ariane", "riscv";
+      riscv,isa = "rv64imafdc";
+      mmu-type = "riscv,sv39";
+      tlb-split;
+      // HLIC - hart local interrupt controller
+      CPU0_intc: interrupt-controller {
+        #interrupt-cells = <1>;
+        interrupt-controller;
+        compatible = "riscv,cpu-intc";
+      };
+    };
+  };
+  memory@80000000 {
+    device_type = "memory";
+    reg = <0x0 0x80000000 0x0 0x40000000>;
+  };
+  leds {
+    compatible = "gpio-leds";
+    heartbeat-led {
+      gpios = <&xlnx_gpio 1 0>;
+      linux,default-trigger = "heartbeat";
+      retain-state-suspended;
+    };
+  };
+  L26: soc {
+    #address-cells = <2>;
+    #size-cells = <2>;
+    compatible = "eth,ariane-bare-soc", "simple-bus";
+    ranges;
+    clint@2000000 {
+      compatible = "riscv,clint0";
+      interrupts-extended = <&CPU0_intc 3 &CPU0_intc 7>;
+      reg = <0x0 0x2000000 0x0 0xc0000>;
+      reg-names = "control";
+    };
+    PLIC0: interrupt-controller@c000000 {
+      #address-cells = <0>;
+      #interrupt-cells = <1>;
+      compatible = "riscv,plic0";
+      interrupt-controller;
+      interrupts-extended = <&CPU0_intc 11 &CPU0_intc 9>;
+      reg = <0x0 0xc000000 0x0 0x4000000>;
+      riscv,max-priority = <7>;
+      riscv,ndev = <30>;
+    };
+//  debug-controller@0 {
+//    compatible = "riscv,debug-013";
+//    interrupts-extended = <&CPU0_intc 65535>;
+//    reg = <0x0 0x0 0x0 0x1000>;
+//    reg-names = "control";
+//  };
+    uart@10000000 {
+      compatible = "ns16750", "ns16550";
+      reg = <0x0 0x10000000 0x0 0x1000>;
+      clock-frequency = <50000000>;
+      current-speed = <115200>;
+      interrupt-parent = <&PLIC0>;
+      interrupts = <1>;
+      reg-shift = <2>; // regs are spaced on 32 bit boundary
+      reg-io-width = <4>; // only 32-bit access are supported
+    };
+    timer@18000000 {
+      compatible = "pulp,apb_timer";
+      interrupts = <0x00000004 0x00000005 0x00000006 0x00000007>;
+      reg = <0x00000000 0x18000000 0x00000000 0x00001000>;
+      interrupt-parent = <&PLIC0>;
+      reg-names = "control";
+    };
+    xps-spi@20000000 {
+      compatible = "xlnx,xps-spi-2.00.b", "xlnx,xps-spi-2.00.a";
+      #address-cells = <1>;
+      #size-cells = <0>;
+      interrupt-parent = <&PLIC0>;
+      interrupts = < 2 2 >;
+      reg = < 0x0 0x20000000 0x0 0x1000 >;
+      fifo-size = <256>;
+      xlnx,family = "kintex7";
+      xlnx,fifo-exist = <0x1>;
+      xlnx,num-ss-bits = <0x1>;
+      xlnx,num-transfer-bits = <0x8>;
+      xlnx,sck-ratio = <0x4>;
+
+      mmc@0 {
+        compatible = "mmc-spi-slot";
+        reg = <0>;
+        spi-max-frequency = <12500000>;
+        voltage-ranges = <3300 3300>;
+        disable-wp;
+      };
+
+      // mmc-slot@0 {
+      //   compatible = "fsl,mpc8323rdb-mmc-slot", "mmc-spi-slot";
+      //   reg = <0>;  //Chip select 0
+      //   spi-max-frequency = <12500000>;
+      //   voltage-ranges = <3300 3300>;
+      //   //interrupts = < 2 2 >;
+      //   //interrupt-parent = <&PLIC0>;
+      // };
+    };
+    eth: lowrisc-eth@30000000 {
+      compatible = "lowrisc-eth";
+      device_type = "network";
+      interrupt-parent = <&PLIC0>;
+      interrupts = <3 0>;
+      local-mac-address = [00 18 3e 02 e3 7f]; // This needs to change if more than one GenesysII on a VLAN
+      reg = <0x0 0x30000000 0x0 0x8000>;
+    };
+    xlnx_gpio: gpio@40000000 {
+      #gpio-cells = <2>;
+      compatible = "xlnx,xps-gpio-1.00.a";
+      gpio-controller ;
+      reg = <0x0 0x40000000 0x0 0x10000 >;
+      xlnx,all-inputs = <0x0>;
+      xlnx,all-inputs-2 = <0x0>;
+      xlnx,dout-default = <0x0>;
+      xlnx,dout-default-2 = <0x0>;
+      xlnx,gpio-width = <0x8>;
+      xlnx,gpio2-width = <0x8>;
+      xlnx,interrupt-present = <0x0>;
+      xlnx,is-dual = <0x1>;
+      xlnx,tri-default = <0xffffffff>;
+      xlnx,tri-default-2 = <0xffffffff>;
+    };
+  };
+};
diff --git a/board/openhwgroup/cva6_genesysII/Kconfig b/board/openhwgroup/cva6_genesysII/Kconfig
new file mode 100644
index 0000000000..42127206a4
--- /dev/null
+++ b/board/openhwgroup/cva6_genesysII/Kconfig
@@ -0,0 +1,27 @@
+if TARGET_OPENHWGROUP_CVA6_GENESYSII
+
+config SYS_BOARD
+	default "cva6_genesysII"
+
+config SYS_VENDOR
+	default "openhwgroup"
+
+config SYS_CPU
+	default "generic"
+
+config SYS_CONFIG_NAME
+	default "openhwgroup_cva6_genesysII"
+
+config SYS_TEXT_BASE
+	default 0x80200000 if SPL
+	default 0x80000000 if !RISCV_SMODE
+	default 0x80200000 if RISCV_SMODE
+
+config SPL_TEXT_BASE
+	default 0x08000000
+
+config BOARD_SPECIFIC_OPTIONS # dummy
+    def_bool y
+    select GENERIC_RISCV
+
+endif
diff --git a/board/openhwgroup/cva6_genesysII/Makefile b/board/openhwgroup/cva6_genesysII/Makefile
new file mode 100644
index 0000000000..b5d4830627
--- /dev/null
+++ b/board/openhwgroup/cva6_genesysII/Makefile
@@ -0,0 +1 @@
+obj-y	+= cva6.o
diff --git a/board/openhwgroup/cva6_genesysII/cva6.c b/board/openhwgroup/cva6_genesysII/cva6.c
new file mode 100644
index 0000000000..f2c558dce7
--- /dev/null
+++ b/board/openhwgroup/cva6_genesysII/cva6.c
@@ -0,0 +1,9 @@
+#include <init.h>
+
+int board_init(void) {
+  return 0;
+}
+
+int ft_board_setup(void *fdt, struct bd_info *bd) {
+  return 0;
+}
diff --git a/common/image-fdt.c b/common/image-fdt.c
index d50e1ba3fe..e1695664ab 100644
--- a/common/image-fdt.c
+++ b/common/image-fdt.c
@@ -203,6 +203,13 @@ int boot_relocate_fdt(struct lmb *lmb, char **of_flat_tree, ulong *of_size)
 		    (void *)(ulong) lmb_alloc_base(lmb, of_len, 0x1000,
 						   env_get_bootm_mapsize()
 						   + env_get_bootm_low());
+		/*
+		 * lmb_alloc_base returns and address that collides with the .bss section. How are we
+		 * supposed to handle this? By adding an lmb->reserved region? How do other boards deal with
+		 * this? How is this even possible???
+		 * Use this hacky workaround for now.
+		 */
+		of_start = (void *)(ulong) 0xbf800000;
 	}
 
 	if (of_start == NULL) {
diff --git a/configs/openhwgroup_cva6_genesysII_defconfig b/configs/openhwgroup_cva6_genesysII_defconfig
new file mode 100644
index 0000000000..82c1f32296
--- /dev/null
+++ b/configs/openhwgroup_cva6_genesysII_defconfig
@@ -0,0 +1,37 @@
+CONFIG_RISCV=y
+CONFIG_DEFAULT_DEVICE_TREE="cva6_genesysII"
+CONFIG_TARGET_OPENHWGROUP_CVA6_GENESYSII=y
+CONFIG_ARCH_RV64I=y
+CONFIG_RISCV_SMODE=y
+CONFIG_OF_BOARD_SETUP=y
+CONFIG_BOOTDELAY=5
+CONFIG_USE_BOOTCOMMAND=y
+CONFIG_BOOTCOMMAND="mmc info; fdt addr $(fdtcontroladdr); fdt set /chosen stdout-path \"/soc/serial@10000000:115200\"; mmc read 90000000 2000 5000; bootm 90000000 - $(fdtcontroladdr)"
+CONFIG_DISPLAY_CPUINFO=y
+CONFIG_SPL_SPI_FLASH_MTD=y
+CONFIG_CMD_GPT=y
+# CONFIG_RANDOM_UUID is not set
+CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
+CONFIG_CMD_PART=y
+CONFIG_OF_EMBED=y
+CONFIG_MMC=y
+# CONFIG_MMC_WRITE is not set
+CONFIG_MMC_SPI=y
+CONFIG_DM_MTD=y
+CONFIG_MTD_SPI_NAND=y
+CONFIG_SPI_FLASH_ATMEL=y
+CONFIG_SPI_FLASH_EON=y
+CONFIG_SPI_FLASH_GIGADEVICE=y
+CONFIG_SPI_FLASH_ISSI=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_SPANSION=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_SPI_FLASH_SST=y
+CONFIG_SPI_FLASH_WINBOND=y
+CONFIG_SPI_FLASH_XMC=y
+# CONFIG_SPI_FLASH_USE_4K_SECTORS is not set
+CONFIG_SPI_FLASH_MTD=y
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_XILINX_SPI=y
diff --git a/include/configs/openhwgroup_cva6_genesysII.h b/include/configs/openhwgroup_cva6_genesysII.h
new file mode 100644
index 0000000000..3a7d2a8af5
--- /dev/null
+++ b/include/configs/openhwgroup_cva6_genesysII.h
@@ -0,0 +1,16 @@
+#ifndef __CONFIG_H
+#define __CONFIG_H
+
+#include <linux/sizes.h>
+
+#define CONFIG_SYS_SDRAM_BASE		0x80000000
+#define CONFIG_SYS_INIT_SP_ADDR		(CONFIG_SYS_SDRAM_BASE + SZ_2M)
+
+#define CONFIG_SYS_LOAD_ADDR		(CONFIG_SYS_SDRAM_BASE + SZ_2M)
+
+#define CONFIG_SYS_MALLOC_LEN       SZ_8M
+#define CONFIG_SYS_BOOTM_LEN        SZ_32M
+
+#define CONFIG_SYS_MAX_FLASH_BANKS 1
+#define CONFIG_SYS_FLASH_BASE 0x0
+#endif
-- 
2.31.1

